xcube/webapi/context.py

```diff
@@ -303,16 +302,12 @@ def get_variable_for_z(self, ds_id: str, var_name: str, z_index: int) -> xr.Data
         return dataset[var_name]
 
     def get_dataset_configs(self) -> List[DatasetConfigDict]:
-        if self._dataset_configs is None:
-            with self._lock:
-                dataset_configs = self._config.get('Datasets', [])
-                dataset_configs += \
-                    self.get_dataset_configs_from_stores()
-                self._dataset_configs = dataset_configs
-                self._maybe_assign_store_instance_ids()
+        with self._lock:
+            if self._dataset_configs is None:
+                self._set_up_data_store_pool_and_dataset_configs()
```

Comment:
Only lock, if needed. Double-check to be on the safe side (see https://en.wikipedia.org/wiki/Double-checked_locking).

```suggestion
        if self._dataset_configs is None:
            with self._lock:
                if self._dataset_configs is None:
                    self._set_up_data_store_pool_and_dataset_configs()
```